File: Práctica3/java/src/deepspace/Hangar.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author daniel
 */

public class Hangar {
    /**
     * Máximo número de escudos y armas (combinados) que puede tenerla estación
     */
    private int maxElements;
    
    /**
     * Array con los shieldboosters que tiene el hangar
     */
    private ArrayList<ShieldBooster> shieldBoosters;
    
    /**
     * Array con las weapons que tiene el hangar
     */
    private ArrayList<Weapon> weapons;
    
    // Constructores
    
    /**
     * Class initializer
     * @param capacity máximo número de escudos y armas combinados
     */
    Hangar(int capacity){
        maxElements=capacity;
        shieldBoosters=new ArrayList<>();
        weapons=new ArrayList<>();
    }
    
    /**
     * Constructor de copia
     * @param h hangar a copiar
     */
    Hangar(Hangar h){
        maxElements=h.maxElements;
        shieldBoosters=new ArrayList<>();
        weapons=new ArrayList<>();
        
        for (ShieldBooster s : h.shieldBoosters)
            addShieldBooster(s);
        for (Weapon w : h.weapons)
            addWeapon(w);
    }
    
    // Getters
    
    /**
     * Getter for maxElements
     * @return maxElements
     */
    public int getMaxElements() {
        return maxElements;
    }
    
    /**
     * Getter for shieldBoosters
     * @return shieldBoosters
     */
    public ArrayList<ShieldBooster> getShieldBoosters() {
        return shieldBoosters;
    }
    
    /**
     * Getter for weapons
     * @return weapons
     */
    public ArrayList<Weapon> getWeapons() {
        return weapons;
    }
    
    /**
     * Comprueba si aún hay espacio para añadir elementos
     * @return true si queda hueco, false en caso contrario
     */
    private boolean spaceAvailable(){
        return maxElements > shieldBoosters.size() + weapons.size();
    }
    
    /**
     * Añade un arma al hangar
     * @param w el arma a añadir
     * @return true si es posible, false en caso contrario
     */
    public boolean addWeapon(Weapon w){
        if (spaceAvailable()){
            return weapons.add(w);
        }else
            return false;
    }
    
    /**
     * Añade un potenciador de escudo
     * @param s escudo a añadir
     * @return true si es posible, false en caso contrario
     */
    public boolean addShieldBooster(ShieldBooster s){
        if (spaceAvailable()){
            return shieldBoosters.add(s);
        }else
            return false;
    }
    
    /**
     * Elimina un arma
     * @param w posición en el array del arma a eliminar
     * @return devuelve el arma extraida si es posible, sino null
     */
    public Weapon removeWeapon(int w){
        if (w >= 0 && w < weapons.size()){
            return weapons.remove(w);
        }else
            return null;
    }
    
    /**
     * Elimina un potenciador de escudo
     * @param s posición en el array del potenciador
     * @return devuelve el potenciador extraido si es posible, sino null
     */
    public ShieldBooster removeShieldBooster(int s){
        if (s >= 0 && s < shieldBoosters.size()){
            return shieldBoosters.remove(s);
        }else
            return null;
    }
    
    // UI version
    
    /**
     * String representation of the object
     * @return string representation
     */
    @Override
    public String toString() {
        String message = "[Hangar] -> Max. elements: " + maxElements
                + ", Shields: " + shieldBoosters.toString()
                + ", Weapons: " + weapons.toString();
        return message;
    }
    
    /**
     * To UI
     */
    HangarToUI getUIversion() {
        return new HangarToUI(this);
    }
}

File: Práctica3/lib/Hangar.rb
#encoding:utf-8

require_relative 'Weapon'
require_relative 'ShieldBooster'
require_relative 'WeaponType'
require_relative 'HangarToUI'

module Deepspace
  class Hangar

    # Constructores
    # @param capacity [Integer] máximo número de escudos y armas que caben en el hangar
    def initialize(capacity)
      # @!attribute [Integer] espacio del hangar
      @maxElements = capacity

      # @!attribute [Array<ShieldBooster>] array de escudos almacenados en el hangar
      @shieldBoosters = []

      # @!attribute [Array<Weapon>] array de armas almacenada en el hangar
      @weapons = []
    end

    # Constructor de copia
    # @param d [Hangar] instancia a copiar
    def self.newCopy(h)
      copy = new(h.maxElements)

      for shieldBooster in h.shieldBoosters
        copy.addShieldBooster(shieldBooster)
      end

      for weapon in h.weapons
        copy.addWeapon(weapon)
      end

      return copy
    end

    attr_reader :shieldBoosters, :weapons, :maxElements

    # Comprueba si queda espacio en el hangar
    # @return [Boolean] true en caso afirmativo, false en caso contrario
    def spaceAvailable
      return @maxElements > @weapons.length + @shieldBoosters.length
    end

    # Añade un nuevo arma al hangar
    # @param w [Weapon] arma a añadir
    # @return [Boolean] true en caso de que se añada correctamente false en caso contario
    def addWeapon(w)
      if spaceAvailable
        @weapons << w
        return true
      else
        return false
      end
    end

    # Elimina un arma del hangar
    # @param w [Integer] posicion del arma a eliminar
    # @return [Weapon] arma eliminada. Si no se encuentra el arma devuelve nil
    def removeWeapon(w)
      if w < @weapons.length && w >= 0
        return @weapons.delete_at(w)
      else
        return nil
      end
    end

    # Añade un nuevo potenciador de escudo al hangar
    # @param w [ShieldBooster] potenciador a ser añadido
    # @return [Boolean] true en caso de que se añada con exito, false en caso contrario
    def addShieldBooster(s)
      if spaceAvailable
        @shieldBoosters << s
        return true
      else
        return false
      end
    end

    # Elimina un escudo del hangar
    # @param s [Integer] posicion en el hangar del escudo a eliminar
    # @return [ShieldBooster] escudo eliminado. Si no es encuentra devuelve nil
    def removeShieldBooster(s)
      if s < @shieldBoosters.length && s >= 0
        return @shieldBoosters.delete_at(s)
      else
        return nil
      end
    end

    # String representation, UI version
    # ==========================================================================

    # String representation of the object
    # @return [String] string representation
    def to_s
      getUIversion().to_s
    end

    # To UI
    def getUIversion
      return HangarToUI.new(self)
    end
  end
end  

